src/NVMLBoardUtilizationAgent.cpp

```diff
@@ -0,0 +1,294 @@
+    {
+        assert(in_policy.size() == M_NUM_POLICY);
+        m_do_write_batch = false;
+
+        double utilization_accelerator = m_signal_available.at("NVML::UTILIZATION_ACCELERATOR").m_last_signal;
+        double accel_freq_request = NAN;
+        double xeon_freq_request = NAN;
+
+        if (!std::isnan(utilization_accelerator)) {
+            if(utilization_accelerator <= in_policy[M_POLICY_ACCELERATOR_UTIL_THRESH_0]) {
+                accel_freq_request = in_policy[M_POLICY_ACCELERATOR_FREQ_SUB_THRESH_0];
+                xeon_freq_request = in_policy[M_POLICY_XEON_FREQ_SUB_THRESH_0];
+            } else if(utilization_accelerator <= in_policy[M_POLICY_ACCELERATOR_UTIL_THRESH_1]) {
+                accel_freq_request = in_policy[M_POLICY_ACCELERATOR_FREQ_SUB_THRESH_1];
+                xeon_freq_request = in_policy[M_POLICY_XEON_FREQ_SUB_THRESH_1];
+            } else {
+                accel_freq_request = in_policy[M_POLICY_ACCELERATOR_FREQ_ABOVE_THRESH_1];
+                xeon_freq_request = in_policy[M_POLICY_XEON_FREQ_ABOVE_THRESH_1];
+            }
+        }
+
+        if (!std::isnan(accel_freq_request) && !std::isnan(xeon_freq_request)) {
+            if (accel_freq_request != m_signal_available.at("NVML::FREQUENCY").m_last_signal ||
+                xeon_freq_request != m_signal_available.at("FREQUENCY").m_last_signal) {
+                // set NVML frequency control
+                auto freq_ctl_itr = m_control_available.find("NVML::FREQUENCY_CONTROL");
+                if (accel_freq_request != freq_ctl_itr->second.m_last_setting) {
+                    m_platform_io.adjust(freq_ctl_itr->second.m_batch_idx, accel_freq_request);
+                    freq_ctl_itr->second.m_last_setting = accel_freq_request;
+                    ++m_accelerator_frequency_requests;
+                }
+
+                // set Xeon frequency control
+                freq_ctl_itr = m_control_available.find("FREQUENCY");
+                if (xeon_freq_request != freq_ctl_itr->second.m_last_setting) {
+                    m_platform_io.adjust(freq_ctl_itr->second.m_batch_idx, xeon_freq_request);
+                    freq_ctl_itr->second.m_last_setting = xeon_freq_request;
+                }
+
+                m_do_write_batch = true;
+            }
+        }
+    }
+
+    //If new values have been adjusted, write
+    bool NVMLBoardUtilizationAgent::do_write_batch(void) const
+    {
+        return m_do_write_batch;
+    }
+
+    // Read signals from the platform and calculate samples to be sent up
+    void NVMLBoardUtilizationAgent::sample_platform(std::vector<double> &out_sample)
+    {
+        assert(out_sample.size() == M_NUM_SAMPLE);
+        // Collect latest signal values
+        for (auto &sv : m_signal_available) {
+            sv.second.m_last_signal = m_platform_io.sample(sv.second.m_batch_idx);
+        }
+    }
+
+    // Wait for the remaining cycle time to keep Controller loop cadence at 1 second
```

Comment:
50ms?

Reply:
is this agent a candidate (guinea pig) for including control loop cadence as a policy value?  See #576 and #1065